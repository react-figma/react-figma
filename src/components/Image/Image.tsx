import * as React from 'react';
import { RectangleProps, Rectangle } from '../rectangle/Rectangle';
import { ResizeMode } from '../../styleTransformers/transformGeometryStyleProperties';
import { StyleSheet, Svg } from '../..';
import { useSvgText } from '../../hooks/useSvgText';

export interface ImageProps extends RectangleProps {
    source: string | { uri: string } | { default: string };
    resizeMode?: ResizeMode;
}

const Image: React.FC<ImageProps> = props => {
    const { style, source, resizeMode } = props;
    let svgText;

    if (typeof source === 'string' && source.slice(0, 4) === '<svg') {
        svgText = source;
    }
    if (typeof source !== 'string' && 'uri' in source && source.uri.slice(-4) === '.svg') {
        svgText = useSvgText(source.uri);
    }
    if (svgText) {
        return <Svg style={style} source={svgText} />;
    }

    return (
        <Rectangle
            {...props}
            style={[StyleSheet.flatten(style), { backgroundImage: source, backgroundSize: resizeMode }]}
        />
    );
};

export { Image };
