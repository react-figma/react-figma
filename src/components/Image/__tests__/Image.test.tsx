import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Image } from '../Image';
import { createFigma } from 'figma-api-stub';

describe('<Image />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('Image with source and layout style properties', () => {
        const tree = renderer
            .create(
                <Image
                    source="./image.png"
                    style={{
                        width: 100,
                        height: 200
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Image with resizeMode prop', () => {
        const tree = renderer
            .create(
                <Image
                    source="./image.png"
                    resizeMode="stretch"
                    style={{
                        width: 100,
                        height: 200
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Image with local SVG', () => {
        const tree = renderer
            .create(
                <Image
                    source="'./logo.svg'"
                    style={{
                        width: 100,
                        height: 100
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Image with remote SVG', () => {
        const tree = renderer
            .create(
                <Image
                    source={{
                        uri:
                            'https://raw.githubusercontent.com/react-figma/react-figma/dfa655881c7944a8ec76ef9758c92a480ae27959/logo.svg'
                    }}
                    style={{
                        width: 100,
                        height: 100
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
