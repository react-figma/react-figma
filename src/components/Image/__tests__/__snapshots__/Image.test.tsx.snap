// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> Image with local SVG 1`] = `
<rectangle
  height={100}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  source="'./logo.svg'"
  style={
    Object {
      "backgroundImage": "'./logo.svg'",
      "backgroundSize": undefined,
      "height": 100,
      "width": 100,
    }
  }
  width={100}
/>
`;

exports[`<Image /> Image with remote SVG 1`] = `
<rectangle
  height={100}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  source={
    Object {
      "uri": "https://raw.githubusercontent.com/react-figma/react-figma/dfa655881c7944a8ec76ef9758c92a480ae27959/logo.svg",
    }
  }
  style={
    Object {
      "backgroundImage": Object {
        "uri": "https://raw.githubusercontent.com/react-figma/react-figma/dfa655881c7944a8ec76ef9758c92a480ae27959/logo.svg",
      },
      "backgroundSize": undefined,
      "height": 100,
      "width": 100,
    }
  }
  width={100}
/>
`;

exports[`<Image /> Image with resizeMode prop 1`] = `
<rectangle
  height={200}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  resizeMode="stretch"
  source="./image.png"
  style={
    Object {
      "backgroundImage": "./image.png",
      "backgroundSize": "stretch",
      "height": 200,
      "width": 100,
    }
  }
  width={100}
/>
`;

exports[`<Image /> Image with source and layout style properties 1`] = `
<rectangle
  height={200}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  source="./image.png"
  style={
    Object {
      "backgroundImage": "./image.png",
      "backgroundSize": undefined,
      "height": 200,
      "width": 100,
    }
  }
  width={100}
/>
`;
